### ISSUE TITLE
[AVM Module Issue]: Cannot use this module if the resource group does not yet exist

### ISSUE THREAD

**Issue by SvenAelterman** [Azure/terraform-azurerm-avm-res-network-routetable#22]
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### (Optional) Module Version

0.2.0

### Description

When defining the resource where the route table is to be created, but that resource group does not yet exist, `terraform plan` and `apply` error with:

```hcl
│ Error: Error: Resource Group "identity-networking-demo-rg-cnc-01" was not found
│
│   with module.udr_identity.data.azurerm_resource_group.parent,
│   on .terraform\modules\udr_identity\main.tf line 1, in data "azurerm_resource_group" "parent":
│    1: data "azurerm_resource_group" "parent" {
```

Relevant snip:

```
resource "azurerm_resource_group" "rg_identity_nw" {
  provider = azurerm.identity

  name     = "identity-networking-demo-rg-cnc-01"
  location = var.default_location

  tags = {
    date-created = "2024-07-21"
  }
}

# Create a UDR to the firewall
module "udr_identity" {
  source  = "Azure/avm-res-network-routetable/azurerm"
  version = "~> 0.2.0"

  name                = "addc-demo-rt-cnc-01"
  location            = var.default_location
  resource_group_name = azurerm_resource_group.rg_identity_nw.name
  tags                = { "date-created" : "2024-07-21" }

  enable_telemetry = try(!local.archetypes.disable_telemetry, false)

  routes = {
    "to-firewall" = {
      name           = "to-firewall"
      address_prefix = "0.0.0.0/0"
      next_hop_type  = "VirtualAppliance"
      # Assumption: Only one firewall (single hub in a single region only)
      next_hop_in_ip_address = module.hubnetworking[0].firewalls.primary.private_ip_address
    }
  }

  providers = {
    azurerm = azurerm.identity
  }
}
```

**Comment by adammontlake** [comment]
Hi @SvenAelterman!
This is expected behavior.
From the [Terraform docs](https://developer.hashicorp.com/terraform/language/data-sources#data-resource-behavior), data resource is evaluated during the plan time unless one of the following conditions prevent it:

- At least one of the given arguments is a managed resource attribute or other value that Terraform cannot predict until the apply step.
- The data resource depends directly on a managed resource that itself has planned changes in the current plan.
- The data resource has custom conditions and it depends directly or indirectly on a managed resource that itself has planned changes in the current plan.

One way to solve this would be to add a depends on to the module call:
`depends_on = [ azurerm_resource_group.rg_identity_nw ]`

**Comment by adammontlake** [comment]
Hi @SvenAelterman!
While this is not a bug, it has made me reconsider the usefulness of the data resource, and why it is even there.
Historically in v0.1 we determined the resource location from the resource group, but since this is no longer the case, I think we no longer require the resource group as a data source and can reference the resource group directly by the var.resource_group_name.

I will mark this as an enhancement and add it to the next version.

Thanks for the feedback!

**Comment by SvenAelterman** [comment]
> Hi @SvenAelterman! This is expected behavior. From the [Terraform docs](https://developer.hashicorp.com/terraform/language/data-sources#data-resource-behavior), data resource is evaluated during the plan time unless one of the following conditions prevent it:
> 
> * At least one of the given arguments is a managed resource attribute or other value that Terraform cannot predict until the apply step.
> * The data resource depends directly on a managed resource that itself has planned changes in the current plan.
> * The data resource has custom conditions and it depends directly or indirectly on a managed resource that itself has planned changes in the current plan.
> 
> One way to solve this would be to add a depends on to the module call: `depends_on = [ azurerm_resource_group.rg_identity_nw ]`

I saw your later comment too, but I have a question about this suggestion about using an explicit dependency. I am already creating an implicit dependency between the UDR and the RG by using the RG's name in the UDR's `resource_group_name` property.

Would an explicit dependency make a difference?

**Comment by adammontlake** [comment]
Hey @SvenAelterman
Absolutely, my interpretation of why this works is that the new resource group has not yet been created, therefore it is not a managed resource by Terraform. So, you are not actually creating a dependency at all, and the data resource is trying to evaluate during the plan time.

However, if you want a real explanation on why this is the behavior in Terraform, I suggest a discussing on the Hashicorp forum ;)

**Comment by adammontlake** [comment]
Hey @SvenAelterman 
The removal of the data resource is merged and new release 0.2.1 is now out.

Thanks for the useful feedback!
